src/Language/Haskell/LSP/VFS.hs

```diff
@@ -148,26 +162,27 @@ changeFromServerVFS initVfs (J.RequestMessage _ _ _ params) = do
 
 -- ---------------------------------------------------------------------
 virtualFileName :: FilePath -> J.NormalizedUri -> VirtualFile -> FilePath
-virtualFileName prefix uri (VirtualFile ver _) =
-  prefix </> show (hash (J.fromNormalizedUri uri)) ++ "-" ++ show ver ++ ".hs"
+virtualFileName prefix uri (VirtualFile _ file_ver _) =
+  prefix </> show (hash (J.fromNormalizedUri uri)) ++ "-" ++ show file_ver ++ ".hs"
 
-persistFileVFS :: VFS -> J.NormalizedUri -> (FilePath, IO ())
+-- | Write a virtual file to a temporary file if it exists in the VFS.
+persistFileVFS :: VFS -> J.NormalizedUri -> Maybe (FilePath, IO ())
 persistFileVFS vfs uri =
   case Map.lookup uri (vfsMap vfs) of
-    Nothing -> error ("File not found in VFS: " ++ show uri ++ show vfs)
-    Just vf@(VirtualFile _v txt) ->
+    Nothing -> Nothing
+    Just vf ->
       let tfn = virtualFileName (vfsTempDir vfs) uri vf
           action = do
             exists <- doesFileExist tfn
-            unless exists (writeFile tfn (Rope.toString txt))
-      in (tfn, action)
+            unless exists (writeFile tfn (Rope.toString (_text vf)))
+      in Just (tfn, action)
```

Comment:
Do these files ever get cleaned up? I feel as if this function should be in CPS style with a handler that passes the filepath and deletes it afterwards

Reply:
They all get cleaned up at he end when the main process exits. I suppose they could be cleaned up individually using the refactoring you suggest but there doesn't seem to be much benefit.